Ginkgo: Added PolicyGeneration on Nightly tests

eloycoto opened the pull request:
This issue covers a few test cases for issue #2029, not all yet, but I'll provide more test cases this week. 

Summary: 

- Added a new policy generation tool under helpers/policygen folder.
- Added exitCode on CMDRes
- Added Kubectl.GetServiceHostPort method.
- Added NightlyK8sPolicies test

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

eloycoto committed:
Ginkgo: Added PolicyGeneration on Nightly tests

- Added a new policy generation tool under helpers/policygen folder.
- Added exitCode on CMDRes
- Added Kubectl.GetServiceHostPort method.
- Added NightlyK8sPolicies test

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

aanm commented on Jenkinstest:
```diff
@@ -0,0 +1,42 @@
+pipeline {
```
Is this file a leftover?

jrajahalme replied:
If this file is added intentionally, it should be added to CODEOWNERS file, as this is the only change in this PR that is pulling in the requirement for Janitor's review.

If this file is here accidentally we may need to add it to .gitignore to avoid it popping up later.

eloycoto replied:
Sorry, I was testing this in my local Jenkins and I did not remove it. My fault!

aanm commented on test/helpers/kubectl.go:
```diff
@@ -67,7 +68,7 @@ func (kub *Kubectl) ExecPodCmd(namespace string, pod string, cmd string) (string
 	stdout := new(bytes.Buffer)
 	stderr := new(bytes.Buffer)
-	exit := kub.Execute(command, stdout, stderr)
+	exit, _ := kub.Execute(command, stdout, stderr)
```
Why are we ignoring the error?

eloycoto replied:
In this case, we only need if the exit was success or not. The error message is not essential.(A lot of them are supposed to fail)

nebril replied:
Still, logging the errors in debug mode may be useful.

eloycoto replied:
The main problem to log things that didn't work cause a lot of noise, and opens tasks like https://github.com/cilium/cilium/issues/2221 

Have a lot of logs that says fail when it is supposed to fail creates a lot of confusion.

aanm commented on test/helpers/kubectl.go:
```diff
@@ -97,7 +98,8 @@ func (kub *Kubectl) GetPods(namespace string, filter string) *CmdRes {
 	filter := "-o jsonpath='{.items[*].metadata.name}'"
-	exit := kub.Execute(
+
+	exit, _ := kub.Execute(
```
same here

eloycoto replied:
In this case, we only need if the exit was success or not. The error message is not essential.(A lot of them are supposed to fail)

nebril replied:
Still, logging the errors in debug mode may be useful.

aanm commented on test/helpers/policygen/actions.go:
```diff
@@ -0,0 +1,69 @@
+	}
+
+	switch exitCode := res.GetExitCode(); exitCode {
+	case 28:
```
Magic numbers? Can you add a comment with the meaning of each number?

joestringer commented on test/helpers/policygen/policygen.go:
```diff
@@ -0,0 +1,133 @@
+		// 	template: map[string]string{
+		// 		"Ports": `[{"protocol": "icmp"}]`,
+		// 	},
+		// },
```
I'm not sure why this is commented, but it would be good to either uncomment&fix, or open an issue and add a line above the comment that links to the GH issue so we can know why this commented code exists, and find out more.

eloycoto committed:
Fix issues with cilium_ds path

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

nebril commented on test/helpers/kubectl.go:
```diff
@@ -114,21 +116,22 @@ func (kub *Kubectl) GetPodNames(namespace string, label string) ([]string, error
+	if err != nil {
+		return "", 0, err
+	}
+	return data.Spec.ClusterIP, int(data.Spec.Ports[0].Port), nil
```
This will panic if `data.Spec.Ports` is empty. Please add a check for it and return an error accordingly.

nebril commented on test/helpers/node.go:
```diff
@@ -77,7 +78,7 @@ func GetVagrantSSHMeta(vmName string) *SSHMeta {
 // the command in the provided buffers. Returns false if the command failed
 // during its execution.
-func (s *SSHMeta) Execute(cmd string, stdout io.Writer, stderr io.Writer) bool {
+func (s *SSHMeta) Execute(cmd string, stdout io.Writer, stderr io.Writer) (bool, error) {
```
I think that at this point returning `bool` is redundant. Can you change it so only `error` is returned?

nebril commented on test/helpers/node.go:
```diff
@@ -95,32 +96,37 @@ func (s *SSHMeta) Execute(cmd string, stdout io.Writer, stderr io.Writer) bool {
 func (s *SSHMeta) ExecWithSudo(cmd string, stdout io.Writer, stderr io.Writer) bool {
 	command := fmt.Sprintf("sudo %s", cmd)
-	return s.Execute(command, stdout, stderr)
+	result, _ := s.Execute(command, stdout, stderr)
```
Returning only error in all `Exec*` methods would make sure that we don't ignore any errors.

nebril commented on test/helpers/policygen/actions.go:
```diff
@@ -0,0 +1,69 @@
+// returns a ResultType struct.
+func HTTPAction(srcPod string, target string, kub *helpers.Kubectl) *ResultType {
+	command := fmt.Sprintf("%s exec -n %s %s -- %s",
+		"kubectl", helpers.DefaultNamespace,
```
I think `"kubectl"` can be moved to first argument of Sprintf, there are enough args here as it is.

nebril commented on test/helpers/policygen/actions.go:
```diff
@@ -0,0 +1,69 @@
+// HTTPAction runs a helpers.CurlFail from specified pod to a specified target.
+// It needs a `helpers.Kubectl` instance to run the comamnd in the pod. It
+// returns a ResultType struct.
+func HTTPAction(srcPod string, target string, kub *helpers.Kubectl) *ResultType {
```
Why do you return a pointer to ResultType? Seems like it's not really necessary to use a pointer here.

nebril commented on test/helpers/policygen/actions.go:
```diff
@@ -0,0 +1,69 @@
+	case 28: //CURLE_OPERATION_TIMEDOUT (28)
+		return &ResultTimeout
+	case 22: //CURLE_HTTP_RETURNED_ERROR
+		return &ResultAuth
```
This is not necessarly true. `CURLE_HTTP_RETURNED_ERROR` return 22 for every http code >= 400. I think we should differentiate between http status codes and parse the output (we can add `--write-out "%{http_code}"` to curl command, so it only outputs the code).

eloycoto replied:
Not sure about this comment. Is our container, in a specific pod, created and handled by us.. why I should add another layer of complexity if we only care that we get a >400.

nebril commented on test/helpers/policygen/actions.go:
```diff
@@ -0,0 +1,69 @@
+	case 22: //CURLE_HTTP_RETURNED_ERROR
+		return &ResultAuth
+	default:
+		log.Infof("HTTPAction returns exitcode '%d' and it's not handle", exitCode)
```
s/handle/handled

nebril commented on test/helpers/policygen/actions.go:
```diff
@@ -0,0 +1,69 @@
+func HTTPActionPrivate(srcPod string, dest *TargetDetails, kub *helpers.Kubectl) *ResultType {
+	return HTTPAction(
+		srcPod,
+		fmt.Sprintf("http://%s/private", dest),
```
This will fail, as `dest` is a struct.

eloycoto replied:
Hi @nebril no, dest has the `String()` method. It's working correctly.

nebril commented on test/helpers/policygen/actions.go:
```diff
@@ -0,0 +1,69 @@
+func HTTPActionPublic(srcPod string, dest *TargetDetails, kub *helpers.Kubectl) *ResultType {
+	return HTTPAction(
+		srcPod,
+		fmt.Sprintf("http://%s/public", dest),
```
This will fail, as `dest` is a struct.

eloycoto replied:
Hi @nebril no, dest has the String() method. It's working correctly.

nebril commented on test/helpers/policygen/actions.go:
```diff
@@ -0,0 +1,69 @@
+}
+
+// HTTPActionPrivate runs a CurlAction to private http target using destTargetDetails
+func HTTPActionPrivate(srcPod string, dest *TargetDetails, kub *helpers.Kubectl) *ResultType {
```
I think there is no need to send pointer to `TargetDetails` here. Copying the struct should be enough here.

nebril commented on test/helpers/policygen/actions.go:
```diff
@@ -0,0 +1,69 @@
+}
+
+// HTTPActionPublic runs a CurlAction to public http target using destTargetDetails
+func HTTPActionPublic(srcPod string, dest *TargetDetails, kub *helpers.Kubectl) *ResultType {
```
I think there is no need to send pointer to `TargetDetails` here. Copying the struct should be enough here.

nebril commented on test/helpers/policygen/actions.go:
```diff
@@ -0,0 +1,69 @@
+
+// PingAction executes a ping from the `srcPod` to the destination Target using
+// Kubectl object. It will return a ResultType based on the exitCode
+func PingAction(srcPod string, dest *TargetDetails, kub *helpers.Kubectl) *ResultType {
```
Why send TargetDetails, when only IP is needed for executing `ping`?

eloycoto replied:
All the actions are called with the same params on models on `ExecTest` so it's like an interface.

nebril commented on test/helpers/policygen/actions.go:
```diff
@@ -0,0 +1,69 @@
+// Kubectl object. It will return a ResultType based on the exitCode
+func PingAction(srcPod string, dest *TargetDetails, kub *helpers.Kubectl) *ResultType {
+	command := fmt.Sprintf("%s exec -n %s %s -- %s",
+		"kubectl", helpers.DefaultNamespace,
```
`"kubectl"` should be a part of first argument to Sprintf.

eloycoto replied:
My fault, should be helpers.Kubectl

eloycoto committed:
Ginkgo: Fix issues with ingress/egress rules

Fix an issue when l3 was ResultTimeout in ingress and Egress has a deny
policy. Egress rules are the first match, so it should check first the
egress part instead ingress

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

eloycoto committed:
Ginkgo: Fix L7 issue

Fix L7 isssues on egress+ingress cases

nebril commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+// Template field is used to render the cilium network policy.
+type PolicyTest struct {
+	name     string
+	kind     string //Egress/ingress
```
Please name the type PolicyTestKind and add Egress/Ingress consts if you want to make sure that values are limited to a certain subset.

nebril commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+
+// GetField method to retrieve the value of any type of the struct.
+// It is used by `TestSpec` to created expected results
+func (conn *ConnTestSpec) GetField(field string) ResultType {
```
I think it would be better to write a `GetMap` method which would return a one-time-use map to access fields by string. My proposal of implementation:
```
func (conn *ConnTestSpec) GetMap() map[string]ResultType {
  return map[string]ResultType{
    "http": conn.HTTP,
    "httpprivate": conn.HTTPPrivate,
    (...)
  }
}
```

This way we can iterate over all types easily, access them by string, don't need to use reflection, and still have underlying compile-time type checking for ConnTestSpec. The only nuissance is having GetMap aligned with all ConnTestSpec fields.

aanm replied:
I think it should be a switch statement because the fields are already constants defined in `./const.go`, something like:
```golang
func (conn *ConnTestSpec) GetField(field fieldTypes) ResultType {
   switch field{
   case HTTPPrivate:
    return conn.HTTPrivate
    case ...
  } 
}
```
(and it should also be a `fieldTypes` for `HTTPPrivate`, etc in `const.go`)

nebril commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+// templates. The result will be stored in the result parameter, the spec
+// parameters is needed to retrieve the source and destination pods and pass
+// the information to the go template.
+func (pol *PolicyTest) SetTemplate(result *map[string]interface{}, spec *TestSpec) error {
```
Maps are passed to functions by reference, no need to pass a pointer.

nebril commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+}
+
+// TargetDetails helper struct to store the IP and Port
+type TargetDetails struct {
```
Please consider using an alias for https://golang.org/pkg/net/#TCPAddr

nebril commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+	l7          PolicyTest
+	SrcPod      string
+	DestPod     string
+	Prefix      string
```
Seems like formatting issue?

nebril commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+
+// GetConnectivyTest returns and array with the expected results of the given
+// connectivy test kind
+func (t *TestSpec) GetConnectivyTest(kind string) []connTestResultType {
```
s/Connectivy/Connectivity
Also please respond to houndci issue.

nebril commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+		connTestResultType{t.l7.kind, t.l7.tests.GetField(kind)}}
+}
+
+// GetTestExpects returns a map with the connTestType and the exepected result
```
s/exepected/expected

nebril commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+	log "github.com/sirupsen/logrus"
+)
+
+// ConnTestSpec Connectivy Test Specification. This is a struct to use in
```
s/Connectivy/Connectivity

nebril commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+	result ResultType
+}
+
+// GetConnectivyTest returns and array with the expected results of the given
```
s/Connectivy/Connectivity

nebril commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+}
+
+// GetConnectivyTest returns and array with the expected results of the given
+// connectivy test kind
```
s/connectivy/connectivity

nebril commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+	return result
+}
+
+// ExecTest runs the connectivyTest for the expected `PolicyTest`. It will
```
s/connectivy/connectivity

nebril commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+// GetTestExpects returns a map with the connTestType and the exepected result
+// based on the `testExpect`
+func (t *TestSpec) GetTestExpects() map[string]ResultType {
+	DetermineStatus := func(testType string) ResultType {
```
Please change name of the variable to `expectedTestResult` or sth like that.

nebril commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+func (t *TestSpec) GetTestExpects() map[string]ResultType {
+	DetermineStatus := func(testType string) ResultType {
+		connTest := t.GetConnectivyTest(testType)
+
```
I don't get why do we need these two for loops. As I understand, `DetermineStatus` returns what result is expected from give testType? What does it matter which rules are matched first?

eloycoto replied:
Easy, this is an example rule:

IngressRule on pod dest:
-- allow PORT 8080 --> ResultTimeout

EgressRule on src pod:
-- only access to /private/ url to port 80:

Testcase:

- If src pod made an http request to  http://destPod:80/private --> ResultTimeout Correct. (The rule that match is  the IngressRule)
- If src pod made an http request to  http://destPod:80/public --> ResultAuth --> The expected was ResultTimeout

So, we need to check the egress  before the ingress part to determine the status of the expected result

nebril commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+	DetermineStatus := func(testType string) ResultType {
+		connTest := t.GetConnectivyTest(testType)
+
+		//First check the egress rules that are the first rules that matchs
```
s/matchs/match

nebril reviewed (changes requested):
Most of my comments are stylistic or nitpicks, but there are some bigger ones. Please ping me for any clarifications.

eloycoto committed:
Ginkgo: helpers.Node refactoring Execute method

Return only the error instead of sucess, err

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

eloycoto committed:
Ginkgo - Fix PR review comments

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

ianvernon commented on test/helpers/kubectl.go:
```diff
@@ -67,8 +68,8 @@ func (kub *Kubectl) ExecPodCmd(namespace string, pod string, cmd string) (string
-	exit := kub.Execute(command, stdout, stderr)
-	if exit == false {
+	err := kub.Execute(command, stdout, stderr)
+	if err != nil {
```
The error should be propagated; change log message on line 75 to be `kub.logger.WithError(err).Errorf(...`

ianvernon commented on test/helpers/node.go:
```diff
@@ -93,34 +94,42 @@ func (s *SSHMeta) Execute(cmd string, stdout io.Writer, stderr io.Writer) bool {
 // ExecWithSudo executes the provided command using sudo privileges via SSH.
 // The stdout  and stderr of the command are written to the specified stdout and
 // stderr buffers accordingly. Returns false if execution of cmd fails.
 func (s *SSHMeta) ExecWithSudo(cmd string, stdout io.Writer, stderr io.Writer) bool {
```
Can you change this function to return an error similar to `Execute` on line 81 above? Would make the functions more uniform in their implementation.

ianvernon commented on test/helpers/policygen/actions.go:
```diff
@@ -0,0 +1,69 @@
+)
+
+// HTTPAction runs a helpers.CurlFail from specified pod to a specified target.
+// It needs a `helpers.Kubectl` instance to run the comamnd in the pod. It
```
comamnd --> command.

ianvernon commented on test/helpers/policygen/actions.go:
```diff
@@ -0,0 +1,69 @@
+	case 22: //CURLE_HTTP_RETURNED_ERROR
+		return ResultAuth
+	default:
+		log.Infof("HTTPAction returns exitcode '%d' and it's not handled", exitCode)
```
log message is a bit verbose. what about `"HTTPAction returned unexpected exit code %d"` ?

ianvernon commented on test/helpers/policygen/actions.go:
```diff
@@ -0,0 +1,69 @@
+		kub)
+}
+
+// NetPerfAction TODO make this function
```
Create a GH issue and ref it here with `GH-XXXX` so that we know this is tracked.

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+	log "github.com/sirupsen/logrus"
+)
+
+// ConnTestSpec Connectivity Test Specification. This is a struct to use in
```
This description isn't clear enough. Specify that it contains the mapping of protocols to an expected result type that is contextual to each test case.

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+)
+
+// ConnTestSpec Connectivity Test Specification. This is a struct to use in
+// PolicyTest to determine the result of the TestCase.
```
There is no test case structure, yet this documentation makes it seem like there is. Change instances of `TestCase` to `test case` in the documentation to eliminate this source of confusion.

ianvernon replied:
Similarly, there is no `PolicyTest` structure ?

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+}
+
+// SetTemplate renders the template field from the PolicyTest struct using go
+// templates. The result will be stored in the result parameter, the spec
```
Add a period after `parameter`.

ianvernon commented on test/helpers/cmd.go:
```diff
@@ -28,18 +28,24 @@ import (
+	stdout   *bytes.Buffer // Stdout from running cmd
+	stderr   *bytes.Buffer // Stderr from running cmd
+	exit     bool          // Whether command successfully executed
+	exitcode int           // The exitcode result from command execution
```
`// The exit code of cmd`

ianvernon commented on test/helpers/kubectl.go:
```diff
@@ -24,6 +24,7 @@ import (
 	"github.com/sirupsen/logrus"
+	"k8s.io/api/core/v1"
 
 	"github.com/cilium/cilium/api/v1/models"
```
Please move this import before the `govalidator` line with an empty line after it:

`
"github.com/cilium/cilium/api/v1/models"

"github.com/asaskevich/govalidator"
...
`

ianvernon commented on test/helpers/kubectl.go:
```diff
@@ -114,21 +116,25 @@ func (kub *Kubectl) GetPodNames(namespace string, label string) ([]string, error
 	return strings.Split(out, " "), nil
 }
 
+// GetServiceHostPort returns the Host and the first Port for the given service name.
```
Host --> host, Port --> port

ianvernon commented on test/helpers/policygen/actions.go:
```diff
@@ -0,0 +1,69 @@
+	"fmt"
+
+	"github.com/cilium/cilium/test/helpers"
+	log "github.com/sirupsen/logrus"
```
Add an empty line between lines 6 and 7.

ianvernon commented on test/helpers/policygen/actions.go:
```diff
@@ -0,0 +1,69 @@
+package policygen
```
Add Cilium copyright header.

ianvernon commented on test/helpers/policygen/actions.go:
```diff
@@ -0,0 +1,69 @@
+	return ResultType{}
+}
+
+// HTTPActionPrivate runs a CurlAction to private http target using destTargetDetails
```
`private http target` --> `/private`

ianvernon commented on test/helpers/policygen/actions.go:
```diff
@@ -0,0 +1,69 @@
+		kub)
+}
+
+// HTTPActionPublic runs a CurlAction to public http target using destTargetDetails
```
`public http target` --> `/public`

ianvernon commented on test/helpers/policygen/actions.go:
```diff
@@ -0,0 +1,69 @@
+	return ResultType{}
+}
+
+// PingAction executes a ping from the `srcPod` to the destination Target using
```
`the destination Target` --> `dest`

ianvernon commented on test/helpers/policygen/actions.go:
```diff
@@ -0,0 +1,69 @@
+}
+
+// PingAction executes a ping from the `srcPod` to the destination Target using
+// Kubectl object. It will return a ResultType based on the exitCode
```
`Returns a ResultType corresponding to the exit code of the ping command.`

ianvernon commented on test/helpers/policygen/const.go:
```diff
@@ -0,0 +1,76 @@
+package policygen
```
Add Cilium copyright header.

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+	return nil
+}
+
+// ResultType each connectivity test needs to have an expected result. This
```
Make sure that when you are writing documentation, to use complete sentences instead of prepending the name of the function / type before the statement of which it follows.

For instance, this can be condensed to:

`ResultType defines the expected result for a connectivity test.`

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+	success bool   // If the cmd exec is valid or not.
+}
+
+// String returns the the ResultType in humman readable format
```
Please be more careful with duplicate words / misspellings:

`// String returns the ResultType in a human-readable format`

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+	return fmt.Sprintf("kind: %s sucess: %t", res.kind, res.success)
+}
+
+// PolicyTestSuite  helper struct to store all different policies types
```
Eliminate the use of phrases like 'helper struct' from documentation; they don't add much value

ianvernon replied:
`// PolicyTestSuite groups together L3, L4, and L7 policy-related tests.`

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+// It needs a described name, the kind of the test (Egrees or Ingress) and the
+// expected result of `ConnTestSpec`
+// Template field is used to render the cilium network policy.
+type PolicyTestKind struct {
```
This name is a bit confusing, as there is a `kind` field within it. What about `PolicyTestMeta` or `PolicyTestMetadata`, or just `PolicyTest`?

eloycoto replied:
@nebril @ianvernon It was PolicyTest and I changed based on @nebril comment. `PolicyTestKind` make more sense for me, I don't really like `PolicyTestMeta`. Could we agree on this?

ianvernon replied:
Sure; this is not a blocking issue, more of a subjective stylistic decision.

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+	l7Checks []PolicyTestKind
+}
+
+// Target struct to define the destination that wee need to use for the
```
Eliminate the use of the word struct in documentation.

ianvernon replied:
defines a destination for traffic when running tests.

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+// Target struct to define the destination that wee need to use for the
+// testcase
+type Target struct {
+	Kind       string // serviceL3, serviceL4, NodePort, Direct
```
I'd prefer it this was made a type. Then, there could be constants of this type representing serviceL3, serviceL4, NodePort, direct, etc.

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+	PortNumber int
+}
+
+// SetPortNumber returns a non used host(Kubernetes node) port to set in the
```
a nonused --> an unused

ianvernon replied:
an unused port on the host to use in a Kubernetes NodePort service.

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+	return t.PortNumber
+}
+
+// GetTarget returns a `TargetDetails` struct with the IP and Port to  be able
```
... and port to run the tests in spec.

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+
+// GetTarget returns a `TargetDetails` struct with the IP and Port to  be able
+// to execute the actions for the test spec. It needs the `TestSpec` parameter
+// to be able to retrieve the service name. It'll return an error if the
```
It'll return an error --> Returns an error if

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+// to execute the actions for the test spec. It needs the `TestSpec` parameter
+// to be able to retrieve the service name. It'll return an error if the
+// service is not defined or cannot be retrieved. This function only returns
+// the first port mapped in the service, it'll not work with multiple ports.
```
service; it will not work with multiple ports.

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+	return nil, fmt.Errorf("Not Implemented yet")
+}
+
+// GetServiceName returns an string with the service name using the spec
```
returns the prefix of spec prefixed with the kind of the the target.

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+	return fmt.Sprintf("%s-%s", strings.ToLower(t.Kind), spec.Prefix)
+}
+
+//GetManifestName returns the manifest filename for the target using the spec
```
Add spaces after `//`

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+}
+
+//GetManifestName returns the manifest filename for the target using the spec
+//parameter
```
Add spaces after `//`

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+	return fmt.Sprintf("%s/%s", helpers.BasePath, t.GetManifestName(spec))
+}
+
+// CreateApplyManifest it creates the manifest for the type of the target and
```
it creates --> creates

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+}
+
+// CreateApplyManifest it creates the manifest for the type of the target and
+// applied it in kubernetes. It will fail if the service manifest cannot be
```
applied --> applies

ianvernon replied:
kubernetes --> Kubernetes.

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+
+// CreateApplyManifest it creates the manifest for the type of the target and
+// applied it in kubernetes. It will fail if the service manifest cannot be
+// created correctly or applied to kubernetes
```
kubernetes --> Kubernetes

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+	return net.JoinHostPort(target.IP, fmt.Sprintf("%d", target.Port))
+}
+
+// TestSpec struct is where a new test specification is defined. It contains
```
`// TestSpec defines a new test specification.`

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+}
+
+// TestSpec struct is where a new test specification is defined. It contains
+// three different rules (l3,l4,l7) and a destination and source pod in which
```
add a space after commas.

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+
+// TestSpec struct is where a new test specification is defined. It contains
+// three different rules (l3,l4,l7) and a destination and source pod in which
+// test will run. Each testSpec has a prefix, that is a label that is used to
```
that is --> which is

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+// TestSpec struct is where a new test specification is defined. It contains
+// three different rules (l3,l4,l7) and a destination and source pod in which
+// test will run. Each testSpec has a prefix, that is a label that is used to
+// group all resources created by the TestSpec. Each test is going to be
```
is going to be --> is

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+	Kub         *helpers.Kubectl
+}
+
+// String: return the testSpec definition on human readable format
```
`// String returns the TestSpec in a human-readable format`

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+		t.l3.name, t.l4.name, t.l7.name, t.Destination.Kind)
+}
+
+// RunTest is the method that runs all the `TestSpec` methods and makes the
```
remove phrases like "is the method" - it's already a function, so we know it is a method / function.

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+
+// RunTest is the method that runs all the `TestSpec` methods and makes the
+// needed assertions for Ginkgo tests. This method will create pods manifest,
+// wait for pods to be ready, apply a new cilium network policy and create a
```
CiliumNetworkPolicy ?

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+// RunTest is the method that runs all the `TestSpec` methods and makes the
+// needed assertions for Ginkgo tests. This method will create pods manifest,
+// wait for pods to be ready, apply a new cilium network policy and create a
+// new Destination (Service, NodePort) if it's needed. When all of this happen
```
if it's needed --> if needed

ianvernon replied:
Then, it will execute the `connectivityTest` and compare the results with the expected results within the test specification.

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+}
+
+// RunTest is the method that runs all the `TestSpec` methods and makes the
+// needed assertions for Ginkgo tests. This method will create pods manifest,
```
create pods manifest --> create pods

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+	gomega.Expect(err).To(gomega.BeNil(), "cannot execute test for %s", t.Prefix)
+}
+
+// Destroy will delete all the pods, cnp and Destinations that was created by
```
deletes the pods, CiliumNetworkPolicies, and Destinations created by `TestSpec` after the specified delay.

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+
+// Destroy will delete all the pods, cnp and Destinations that was created by
+// `TestSpec`. It needs a delay parameter that means that it'll delete it after
+// the specified time. (This is to keep the Nightly test under consider load)
```
What does the last sentence mean? What is "consider load"?

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+	}
+}
+
+// GetManifestName returns an string with the `TestSpec` manifest name
```
returns a string

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+	return fmt.Sprintf("%s/%s", helpers.BasePath, t.GetManifestName())
+}
+
+// CreateManifests creates a new pod manifest. It will set a random prefix for
```
Use present tense in documentation; `It sets a random prefix...`

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+}
+
+// CreateManifests creates a new pod manifest. It will set a random prefix for
+// the `TestCase` and it creates two new pods (srcPod and DestPod). It will
```
and it creates --> and creates

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+
+// CreateManifests creates a new pod manifest. It will set a random prefix for
+// the `TestCase` and it creates two new pods (srcPod and DestPod). It will
+// return an error in case that the manifest cannot be created.
```
"Returns an error if the manifest cannot be created"

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+	return nil
+}
+
+// ApplyManifest applies a new deployment manifest into the kubernetes cluster.
```
kubernetes --> Kubernetes

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+}
+
+// ApplyManifest applies a new deployment manifest into the kubernetes cluster.
+// It'll return an error if the manifest cannot be applied correctly
```
Returns an error if...

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+}
+
+// GetPodMetadata returns a map with the pod name and the IP for the pods used
+// by the `TestSpec`. It will return an error in case that the pod info cannot
```
Use present tense in documentation: Returns an error if the pod info cannot be retrieved correctly

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+	return result, nil
+}
+
+// NetworkPolicyCreate returns a network policy based on the `TestSpec` l3,l4
```
network policy --> CiliumNetworkPolicy

ianvernon replied:
Add spaces after commas.

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+}
+
+// NetworkPolicyCreate returns a network policy based on the `TestSpec` l3,l4
+// and l7 rules. It will return an error if any of the `PolicyTest` set
```
Returns an error if...

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+	return fmt.Sprintf("%s_policy.json", t.Prefix)
+}
+
+// NetworkPolicyApply applies the cilium network policy in kubernetes and wait
```
cilium network policy --> CiliumNetworkPolicy

ianvernon replied:
kubernetes --> Kubernetes

ianvernon replied:
... waits until the statuses of all pods have been updated. Returns an error if the status of the pods did not update or if the policy was unable to be applied.

ianvernon commented on test/helpers/policygen/policygen.go:
```diff
@@ -0,0 +1,109 @@
+package policygen
```
Add Cilium copyright header.

ianvernon commented on test/helpers/policygen/policygen.go:
```diff
@@ -0,0 +1,109 @@
+	},
+}
+
+// GeneratedTestSpec retuns a `TestSpec` array with all the policies
```
what defines a policy possibility?

ianvernon commented on test/helpers/kubectl.go:
```diff
@@ -97,11 +98,12 @@ func (kub *Kubectl) GetPods(namespace string, filter string) *CmdRes {
 		stdout, nil)
 
-	if exit == false {
+	if err != nil {
```
If the error is meaningful, please put it as part of the `fmt.Errorf` statement on line 107-108.

ianvernon commented on test/helpers/policygen/actions.go:
```diff
@@ -0,0 +1,69 @@
+// Kubectl object. It will return a ResultType based on the exitCode
+func PingAction(srcPod string, dest TargetDetails, kub *helpers.Kubectl) ResultType {
+	command := fmt.Sprintf("%s exec -n %s %s -- %s",
+		"kubectl", helpers.DefaultNamespace,
```
Make the `kubectl` in the `test/helpers` package public and use it here instead of hardcoding this string.

ianvernon commented on test/helpers/policygen/actions.go:
```diff
@@ -0,0 +1,69 @@
+// returns a ResultType struct.
+func HTTPAction(srcPod string, target string, kub *helpers.Kubectl) ResultType {
+	command := fmt.Sprintf("%s exec -n %s %s -- %s",
+		"kubectl", helpers.DefaultNamespace,
```
Make the `kubectl` in the `test/helpers` package public and use it here instead of hardcoding this string.

ianvernon commented on test/helpers/policygen/actions.go:
```diff
@@ -0,0 +1,69 @@
+	if res.WasSuccessful() {
+		return ResultOK
+	}
+	return ResultTimeout
```
Will the result always be a result timeout if the result was not successful?

eloycoto replied:
Yes

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+			IP:   vals[0],
+		}, nil
+	}
+	return nil, fmt.Errorf("Not Implemented yet")
```
`fmt.Errorf("%s not implemented yet", t.Kind)`

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+// NetworkPolicyCreate returns a network policy based on the `TestSpec` l3,l4
+// and l7 rules. It will return an error if any of the `PolicyTest` set
+// Template fails or if spec cannot be dump as string
+func (t *TestSpec) NetworkPolicyCreate() (string, error) {
```
NetworkPolicyCreate --> CreateCiliumNetworkPolicy

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+	result ResultType
+}
+
+// getConnectivityTest returns and array with the expected results of the given
```
and --> an

ianvernon commented on test/helpers/policygen/models.go:
```diff
@@ -0,0 +1,664 @@
+				}
+			}
+		}
+		// If no resulttype for egress, we need to check if any specific
```
Use consistency in comments esp. when referring to types. Change resutltType and resulttype --> ResultType

eloycoto committed:
Ginkgo: PolicyGen fix typos

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

ianvernon reviewed (approved):
Comments are fixed well-enough that I think this can be merged
